File: problems/2183_count-array-pairs-divisible-by-k/solution.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var countPairs = function (nums, k) {
    const gcd = (a, b) => (b === 0 ? a : gcd(b, a % b));
    const counts = new Map();
    for (const num of nums) {
        const g = gcd(num, k);
        counts.set(g, (counts.get(g) || 0) + 1);
    }

    let total = 0;
    const gs = [...counts.keys()];
    for (let i = 0; i < gs.length; i++) {
        for (let j = i; j < gs.length; j++) {
            if ((gs[i] * gs[j]) % k !== 0) {
                continue;
            }
            if (i === j) {
                const c = counts.get(gs[i]);
                total += (c * (c - 1)) / 2;
            } else {
                total += counts.get(gs[i]) * counts.get(gs[j]);
            }
        }
    }
    return total;
};

File: problems/2183_count-array-pairs-divisible-by-k/solution.cpp
class Solution {
  public:
    long long countPairs(vector<int> &nums, int k) {
        unordered_map<long long, long long> counts;
        for (int num : nums) {
            counts[gcd(num, k)]++;
        }

        long long total = 0;
        vector<long long> gs;
        gs.reserve(counts.size());
        for (const auto &[g, c] : counts) {
            gs.push_back(g);
        }
        for (size_t i = 0; i < gs.size(); i++) {
            for (size_t j = i; j < gs.size(); j++) {
                if (gs[i] * gs[j] % k != 0) {
                    continue;
                }
                if (i == j) {
                    long long c = counts[gs[i]];
                    total += c * (c - 1) / 2;
                } else {
                    total += counts[gs[i]] * counts[gs[j]];
                }
            }
        }
        return total;
    }
};
